datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachFlowConfig.java

```diff
@@ -1,40 +1,79 @@
+import static com.sequenceiq.datalake.flow.detach.SdxDetachEvent.SDX_DETACH_STACK_FAILED_EVENT;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachEvent.SDX_DETACH_STACK_SUCCESS_EVENT;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachEvent.SDX_DETACH_STACK_SUCCESS_WITH_EXTERNAL_DB_EVENT;
 import static com.sequenceiq.datalake.flow.detach.SdxDetachState.FINAL_STATE;
 import static com.sequenceiq.datalake.flow.detach.SdxDetachState.INIT_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_ATTACH_NEW_CLUSTER_FAILED_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_ATTACH_NEW_CLUSTER_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_CLUSTER_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_EXTERNAL_DB_FAILED_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_EXTERNAL_DB_STATE;
 import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_FAILED_STATE;
-import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_FINISHED_STATE;
-import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_IN_PROGRESS_STATE;
-import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_START_STATE;
-
-import java.util.List;
-
-import org.springframework.stereotype.Component;
-
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_STACK_FAILED_STATE;
+import static com.sequenceiq.datalake.flow.detach.SdxDetachState.SDX_DETACH_STACK_STATE;
 import com.sequenceiq.flow.core.config.AbstractFlowConfiguration;
 import com.sequenceiq.flow.core.config.RetryableFlowConfiguration;
+import java.util.List;
+import org.springframework.stereotype.Component;
 
 @Component
 public class SdxDetachFlowConfig extends AbstractFlowConfiguration<SdxDetachState, SdxDetachEvent> implements RetryableFlowConfiguration<SdxDetachEvent> {
 
     private static final List<Transition<SdxDetachState, SdxDetachEvent>> TRANSITIONS = new Transition.Builder<SdxDetachState, SdxDetachEvent>()
-            .defaultFailureEvent(SdxDetachEvent.SDX_DETACH_FAILED_EVENT)
+            .defaultFailureEvent(SDX_DETACH_FAILED_EVENT)
+
+            .from(INIT_STATE).to(SDX_DETACH_CLUSTER_STATE)
+            .event(SDX_DETACH_EVENT).noFailureEvent()
+
+            .from(SDX_DETACH_CLUSTER_STATE).to(SDX_DETACH_STACK_STATE)
+            .event(SDX_DETACH_CLUSTER_SUCCESS_EVENT).defaultFailureEvent()
+
+            .from(SDX_DETACH_STACK_STATE).to(SDX_DETACH_EXTERNAL_DB_STATE)
+            .event(SDX_DETACH_STACK_SUCCESS_WITH_EXTERNAL_DB_EVENT)
+            .failureState(SDX_DETACH_STACK_FAILED_STATE)
+            .failureEvent(SDX_DETACH_STACK_FAILED_EVENT)
+
+            .from(SDX_DETACH_STACK_STATE).to(SDX_ATTACH_NEW_CLUSTER_STATE)
+            .event(SDX_DETACH_STACK_SUCCESS_EVENT)
+            .failureState(SDX_DETACH_STACK_FAILED_STATE)
```

Comment:
You have state transitions from SDX_DETACH_STACK_STATE. This looks wrong.
I think line number 46 should be
`.from(SDX_DETACH_EXTERNAL_DB_STATE).to(SDX_ATTACH_NEW_CLUSTER_STATE)`

Reply:
I think lines 49 through 52 can be removed.

Reply:
The reason I have it this way is because the SDX_DETACH_STACK_STATE can transition to the EXTERNAL_DB_STATE if the cluster has an external database, but if it doesn't it can go directly to the ATTACH_NEW_CLUSTER_STATE does that make sense?
If you look in SdxDetachActions the action for SDX_DETACH_STACK_STATE will conditionally send the SDX_DETACH_STACK_SUCCESS_WITH_EXTERNAL_DB_EVENT if the cluster has an external DB and the SDX_DETACH_STACK_SUCCESS_EVENT if it doesn't.

This is my first time writing flow steps though so I may be incorrect with how I'm doing this, this just made sense to me.

Reply:
I see.